### DISCUSSION TITLE
Attachments File Templates don't show up in Command Palette - What am I missing?

### DISCUSSION THREAD

**Discussion by supertoothy** [zim-desktop-wiki/zim-desktop-wiki#1822] [Q&A]
The command palette is really useful to do things without breaking flow. I have a few file templates like freeplane mindmaps, orgmode files and treeline databases that I embed in my Zim pages as attachments.

I'd like the Command Palette to show me the file templates available. It only shows me insert > new attachment Empty or New Attachment File Templates.

The only way I can get the Command Palette to show me options of file types is if I first insert it via the menu bar (Insert / New Attachment). Once I have done this, Zim will correctly show me a list of file template options in the Command Palette for the rest of the session. i.e if I quit Zim and restart it, I have to go through the same process to have file templates specifically available in the Command Palette.

What am I missing? Is anyone else able to replicate this?

**Comment by introt** [comment]
Paging @datenwurm, the plugin creator. Sounds like the menu items are cached and not initialized prior to interaction via the menu bar. If so, one initialization on first load would take care of this without slowing down subsequent uses; a restart or manual first browse of the menu after editing templates sounds like a good tradeoff.

**Reply by datenwurm** [reply]
I've looked into the matter and it seems that the ```New Attachment``` menu is only populated when the menu item is clicked. This has to do with the fact that Zim is not actively watching the ```Templates``` directory. 

A quick fix for this would be to add following code to the top of the ```_init_store``` method inside the ```commandpalette.py```:
```
		action = self.window.pageview.actiongroup.get_action('insert_new_file_menu')
		self.window.pageview._update_new_file_submenu(action)
```

Other ways to get the latest content of the ```Templates``` directory are (as @introt pointed out) : 
* populate the menu items manually by clicking ```Insert > New Attachment```, or 
* restart zim

@jaap-karssenberg: What do you think of this? The above source code fix within the commandpalette.py seems a bit hacky to me. I've checked the Zim source code a bit and there seems to be a class called ```FileTreeWatcher```. If I understand the source code correctly it could be used to watch the ```Templates```-directory and auto-populate the ```New Attachment``` menu when files are added/removed?

**Comment by jaap-karssenberg** [comment]
Yes, the clean solution would be to have those actions always instantiated
so the command palette can find them and update them when a change is
detected.

Maybe have a look at the "zim.gui.customtools" module for ideas how to
implement this. There we use a "CustomToolManager" and a "CustomToolUI"
class to manage the folder and also connect actions to the folder content.
However implementation details may be very different, could help to think
along these lines as a pattern.
